[1.20.x] Pull Uncontested changes from MCF/1.20.x

Shadows-of-Fire opened the pull request:
This PR is a trimmed down version of https://github.com/neoforged/NeoForge/pull/14

It includes all changes up through MinecraftForge [47.1.33](https://maven.minecraftforge.net/net/minecraftforge/forge/1.20.1-47.1.33/forge-1.20.1-47.1.33-changelog.txt) with the exception of the following:

1. https://github.com/MinecraftForge/MinecraftForge/pull/9642
2. https://github.com/MinecraftForge/MinecraftForge/pull/9292
3. https://github.com/MinecraftForge/MinecraftForge/pull/9535
  a. Including derivatives of this PR
4. https://github.com/MinecraftForge/MinecraftForge/pull/9428
  a. Clean merging this PR depended on 9535 due to patch changes.
5. https://github.com/MinecraftForge/MinecraftForge/pull/9453
6. https://github.com/MinecraftForge/MinecraftForge/pull/9664
7. https://github.com/MinecraftForge/MinecraftForge/pull/9175

Each PR listed here will be opened separately as to keep record of what happens to these changes.

bconlon1 committed:
AlterGroundEvent for modifying block placement performed by AlterGroundDecorator (#9637)

Meldexun committed:
Add LivingBreatheEvent and LivingDrownEvent (#9525)

ChampionAsh5357 committed:
Particle Description Data Provider.

PaintNinja committed:
Add option to advertise dedicated servers to LAN.

nathanatnorthernbingedotcom committed:
Fix creative mode screen not correctly using CreativeModeTab::getTabsImage (#9627)

DBotThePony committed:
Add PlayerSpawnPhantomsEvent, utilized to block or forcefully allow PhantomSpawner to spawn phantoms (#9644)

pupnewfster committed:
Make item name rendering and status bar rendering respect additional gui overlays being rendered by mods (#9648)

PaintNinja committed:
Support IPv6 address compression for logged IPs

Technici4n committed:
Add `EmptyEnergyStorage` (#9487)

LexManos committed:
Make the FireworkShapes enum extensible (Closes #9486)

SilverDavidMC committed:
New hook in IClientBlockExtensions to handle enabling tinting on breaking particles. #9446

RealMangorage committed:
Fixed AdvancementsScreen.java.patch buttons not rendering (#9649)

Kovak committed:
Fix the Emissive Rendering for Experimental Light Pipeline (#9651)

Random832 committed:
Fix conflicting keybindings not having the correct click count set. #9360

Jonathing committed:
Re-add in-game mod menu (#9652)

brandon3055 committed:
Fixed FluidUtil#tryFillContainer returning invalid result when simulating #9358

JustRed23 committed:
Add LivingSwapHandItemsEvent

Jonathing committed:
Fix parameter names in IForgeDimensionSpecialEffects.adjustLightmapColors to better describe what they are. (#9656)

LegoMaster3650 committed:
Added Action value to PlayerInteractEvent.LeftClickEvent to expose what action fired the event. #9175

LexManos committed:
Cleanup usages of static imports and build script so that our build doesn't spam useless error like messages.
So that it is easier to see real errors.
Add compatibility checking to standard testing tasks.

Shadows-of-Fire commented:
A PR for the `MakeBrainEvent` is being held until merge of this one, as the commit history depends on the changes in this PR.

shartte commented on patches/minecraft/net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.java.patch:
```diff
@@ -0,0 +1,11 @@
+          BlockPos blockpos = p_225975_.m_6630_(i);
+          if (Feature.m_65788_(p_225974_.m_226058_(), blockpos)) {
+-            p_225974_.m_226061_(blockpos, this.f_69303_.m_213972_(p_225974_.m_226067_(), p_225975_));
++            p_225974_.m_226061_(blockpos, net.minecraftforge.event.ForgeEventFactory.alterGround(p_225974_.m_226058_(), p_225974_.m_226067_(), blockpos, this.f_69303_.m_213972_(p_225974_.m_226067_(), p_225975_)));
```
Is it a potential performance problem firing this via event-bus from the inner-loop of a feature during worldgen? I have no actual numbers, though.

shartte commented on src/main/java/net/minecraftforge/event/level/AlterGroundEvent.java:
```diff
@@ -0,0 +1,79 @@
+ * <p>
+ * This event is not {@linkplain Cancelable cancellable}.
+ * <p>
+ * This event is fired on the {@linkplain MinecraftForge#EVENT_BUS main Forge event bus}
+ * only on the {@linkplain net.minecraftforge.fml.LogicalSide#SERVER logical server}.
```
Will this event fire on the worldgen threads, meaning it's invalid to access most of the game-state outside the worldgen chunk here?

Shadows-of-Fire replied:
Yeah, that would be correct. While that's expected for something like this, it should be noted somewhere.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/neoforged/NeoForge?pullRequest=20) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/neoforged/NeoForge?pullRequest=20) before we can accept your contribution.<br/>**10** out of **18** committers have signed the CLA.<br/><br/>:white_check_mark: bconlon1<br/>:white_check_mark: ChampionAsh5357<br/>:white_check_mark: Technici4n<br/>:white_check_mark: pupnewfster<br/>:white_check_mark: RealMangorage<br/>:white_check_mark: Random832<br/>:white_check_mark: Jonathing<br/>:white_check_mark: JustRed23<br/>:white_check_mark: DBotThePony<br/>:white_check_mark: Shadows-of-Fire<br/>:x: Meldexun<br/>:x: PaintNinja<br/>:x: nathanatnorthernbingedotcom<br/>:x: LexManos<br/>:x: Kovak<br/>:x: brandon3055<br/>:x: LegoMaster3650<br/>:x: SilverDavidMC<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/neoforged/NeoForge?pullRequest=20) it.</sub>

Technici4n commented:
I contest https://github.com/MinecraftForge/MinecraftForge/commit/30f255d76c4c2e915414a24e6d8aed42f429817e - see #26.

LexManos committed:
Replace string with forge tag in vanilla recipes. Closes #9062

LexManos committed:
Fix tag removal functionality that broke during the 1.19 update. Closes #9053 and #8949

pupnewfster commented:
@Shadows-of-Fire could you please rebase this PR and also remove the commit that Tech contested as #26 was indeed a more appropriate fix (and has now been merged). I also think it might be worth extracting the alter ground stuff into its own PR as that also seems like it has some questions relating it and how much sense it makes to pull in its current state.

Additionally it may also make sense to grab the following upstream commits:
- https://github.com/MinecraftForge/MinecraftForge/commit/2745531f08bf159e1346f1b69cbe757ed7e8d67b
- https://github.com/MinecraftForge/MinecraftForge/commit/bcdd44aece5271c1f8e2914bacf0ba065be3843b

In regards to the other upstream commits that have not been addressed yet:
- some are not needed due to #26
- one should probably be added to/made as a PR with the earlier brain related stuff. During the mini NAP cpw said we want to have somewhere to point to regarding upstream changes for if we decide that they are not being merged in and for why, so even if it is unlikely the brain things will be merged (especially as is) we still want a PR.
- https://github.com/MinecraftForge/MinecraftForge/commit/881725c7d591bd44197d0cb82903493eb127cb6f should be its own PR as we need to decide on how we want to handle the logical breaking change it introduces as `BuildCreativeModeTabContentsEvent` is documented as only being fired on the logical client... so given it moves the hook that calls this event to expose it as something that people can call server side I think that qualifies as being breaking.

Shadows-of-Fire commented:
Yeah, unfortunately nuking the commit means I have to rewind back to head and reapply
`screm`

pupnewfster commented on src/main/resources/forge.exc:
```diff
@@ -131,4 +132,6 @@ net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.pr
\ No newline at end of file
+net/minecraft/world/level/storage/loot/LootPool.<init>([Lnet/minecraft/world/level/storage/loot/entries/LootPoolEntryContainer;[Lnet/minecraft/world/level/storage/loot/predicates/LootItemCondition;[Lnet/minecraft/world/level/storage/loot/functions/LootItemFunction;Lnet/minecraft/world/level/storage/loot/providers/number/NumberProvider;Lnet/minecraft/world/level/storage/loot/providers/number/Numbe ...
+net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.placeBlockAt(Lnet/minecraft/world/level/levelgen/feature/treedecorators/TreeDecorator$Context;Lnet/minecraft/core/BlockPos;Lnet/minecraftforge/event/level/AlterGroundEvent$StateProvider;)V=|p_225974_,p_225975_,eventProvider
+net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.placeCircle(Lnet/minecraft/world/level/levelgen/feature/treedecorators/TreeDecorator$Context;Lnet/minecraft/core/BlockPos;Lnet/minecraftforge/event/level/AlterGroundEvent$StateProvider;)V=|p_225971_,p_225972_,eventProvider
```
exc entries should be sorted alphabetically. `net/minecraft/world/level/levelgen` comes before `/net/minecraft/world/level/storage`

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -386,6 +370,21 @@
+    }
+ 
+    private void m_21312_() {
+-      ItemStack itemstack = this.m_6844_(EquipmentSlot.OFFHAND);
```
This patch seems to be off as this first line isn't actually changing, please regen patches after running setup and it should fix it

pupnewfster commented on build.gradle:
```diff
@@ -263,6 +268,8 @@ project(':clean') {
         }
     }
+    
+    tasks.named('compileJava') { mustRunAfter 'extractMapped' }
```
Is using `tasks.named` enough to make this properly lazily configured or does this need a `.configure` before the action?

Shadows-of-Fire replied:
No clue, that's a @Matyrobbrt question

Matyrobbrt replied:
[Yes](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/api/internal/DefaultNamedDomainObjectCollection.java#L370), the overload with an action immediately calls `configure` on the provider before returning, so it's as lazy as `named().configure` is.

pupnewfster commented on build.gradle:
```diff
@@ -1340,9 +1348,7 @@ project(':forge') {
-    tasks.withType(JavaCompile).configureEach {
-        options.encoding = 'UTF-8' // Use the UTF-8 charset for Java compilation
-    }
+    tasks.named('compileJava') { mustRunAfter 'extractMapped' }
```
Is using `tasks.named` enough to make this properly lazily configured or does this need a `.configure` before the action?

pupnewfster commented on src/main/java/net/minecraftforge/event/level/AlterGroundEvent.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    @FunctionalInterface
+    public static interface StateProvider
```
Unnecessary `static` modifier

pupnewfster commented on src/main/java/net/minecraftforge/event/level/AlterGroundEvent.java:
```diff
@@ -0,0 +1,118 @@
+        /**
+         * Gets the BlockState that will be placed at the passed position.
+         */
+        public BlockState getState(RandomSource random, BlockPos state);
```
Unnecessary `public` modifier

pupnewfster commented on patches/minecraft/net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.java.patch:
```diff
@@ -0,0 +1,62 @@
+       }
+ 
+       if (!list.isEmpty()) {
++         var eventProvider = net.minecraftforge.event.ForgeEventFactory.alterGround(p_225969_, list, this.f_69303_::m_213972_);
```
Given only the ones that are at the first element's y get placed would it make sense to move this one line down and also pass that `y` value (`i`) into the event? I know it is possible for callers to get the value by querying the passed in list, but then we can more explicitly declare in the docs for `AlterGroundEvent#getPositions` what the `y` level is that matters (and maybe even allow changing it if there is a decent use case for doing so)

Shadows-of-Fire replied:
I think we can shelve that for future development of the event if there is a usecase for messing with the y-level, personally it wouldn't make any sense (the y-level is the detected ground level, so messing with it would just result in transforming something that isn't the ground).

I imagine most people will not need to care what the y-level is, as they will just be supplying a `StateProvider` which will only be invoked on positions matching the correct y-level.

pupnewfster replied:
Fair enough, especially given the change/addition of passing the y value can always be made in a non-breaking way if someone down the road ends up needing it.

pupnewfster commented on patches/minecraft/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -43,31 +43,15 @@
--            boolean flag1 = !this.m_6040_() && !MobEffectUtil.m_19588_(this) && (!flag || !((Player)this).m_150110_().f_35934_);
-+         if (!this.getEyeInFluidType().isAir() && !this.m_9236_().m_8055_(BlockPos.m_274561_(this.m_20185_(), this.m_20188_(), this.m_20189_())).m_60713_(Blocks.f_50628_)) {
-+            boolean flag1 = this.canDrownInFluidType(this.getEyeInFluidType()) && !MobEffectUtil.m_19588_(this) && (!flag || !((Player)this).m_150110_().f_35934_);
++         net.minecraftforge.common.ForgeHooks.onLivingBreathe(this, -m_7302_(0), m_7305_(0));
```
I am not quite sure but I think just passing `0` to both the `decreaseAirSupply` and `increaseAirSupply` methods is incorrect?  For vanilla implementations of these methods it seems like it is probably fine, but if a mod were to override say `decreaseAirSupply` for example and have an entity that's air supply goes down non linearly this would be lost. More likely similar to what vanilla does we should mass `getAirSupply` to each of these methods and then given we care about the difference potentially subtract `getAirSupply` from the result of each. So basically something along the lines of:
```java
int airSupply = this.getAirSupply();
net.minecraftforge.common.ForgeHooks.onLivingBreathe(this, airSupply - decreaseAirSupply(airSupply), increaseAirSupply(airSupply) - airSupply);
```

Shadows-of-Fire replied:
Makes sense, what we really want is the delta between the two values, and since the method is virtual we need to compute the change instead of attempting to be clever with passing `0`.

pupnewfster commented on src/main/java/net/minecraftforge/event/entity/living/LivingBreatheEvent.java:
```diff
@@ -0,0 +1,68 @@
+ * <br>
+ * {@link #canBreathe()} contains if this entity can breathe.<br>
+ * {@link #getConsumeAirAmount()} contains the amount of air that will be consumed if this entity can't breathe.<br>
+ * {@link #getRefillAirAmount()} contains the amount of air that will be refilled if this entity can breathe.<br>
```
These javadocs should be on the methods themselves instead of randomly being up here.

pupnewfster commented on src/main/java/net/minecraftforge/event/entity/living/LivingBreatheEvent.java:
```diff
@@ -0,0 +1,68 @@
+/**
+ * LivingBreatheEvent is fired whenever a living entity ticks.<br>
+ * <br>
+ * This event is fired via the {@link ForgeHooks#onLivingBreathe(LivingEntity, int, int)}.<br>
```
We should mention which bus this fires on.

pupnewfster commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -403,12 +412,16 @@ public static ChatDecorator getServerChatSubmittedDecorator()
-            //   |-----------------|        |-------------------------|  |-------------------------|    |---------| |--|   |---------------|
-            "((?:[a-z0-9]{2,}:\\/\\/)?(?:(?:[0-9]{1,3}\\.){3}[0-9]{1,3}|(?:[-\\w_]{1,}\\.[a-z]{2,}?))(?::[0-9]{1,5})?.*?(?=[!\"\u00A7 \n]|$))",
-            Pattern.CASE_INSENSITIVE);
+        // schema ipv4 OR namespace port path ends
```
Formatting on which bit corresponds to what got absolutely nuked here

pupnewfster commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1521,9 +1549,72 @@ public static <T> HolderLookup.RegistryLookup<T> wrapRegistryLookup(final Holder
+    public static void onLivingBreathe(LivingEntity entity, int consumeAirAmount, int refillAirAmount)
+    {
+        boolean isAir = entity.getEyeInFluidType().isAir() || entity.level().getBlockState(BlockPos.containing(entity.getX(), entity.getEyeY(), entity.getZ())).is(Blocks.BUBBLE_COLUMN);
+        boolean canBreathe = isAir || !entity.canDrownInFluidType(entity.getEyeInFluidType()) || MobEffectUtil.hasWaterBreathing(entity) || (entity instanceof Player && ((Player) entity).getAbilities().invulnerable);
```
This is wrong as it changes the behavior of water breaking potions to increase the air supply when in vanilla they just prevent it from being further decreased. This `canBreathe` boolean in vanilla would effectively be a `dontDecreaseAirSupply` boolean

Shadows-of-Fire replied:
Worse than that, most of the "can breathe in water" flags do not increas the air supply.  The only thing that increases it is being in air.

pupnewfster replied:
Or being in a bubble column

pupnewfster commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1521,9 +1549,72 @@ public static <T> HolderLookup.RegistryLookup<T> wrapRegistryLookup(final Holder
+            }
+        }
+
+        if (isAir && !entity.level().isClientSide && entity.isPassenger() && entity.getVehicle() != null && !entity.getVehicle().canBeRiddenUnderFluidType(entity.getEyeInFluidType(), entity))
```
Pretty sure this is supposed to be `!isAir`

Shadows-of-Fire replied:
You would be correct

pupnewfster commented on src/main/java/net/minecraftforge/event/entity/living/LivingDrownEvent.java:
```diff
@@ -0,0 +1,46 @@
+ * LivingDrownEvent is fired whenever a living entity can't breathe and its air aupply is less than or equal to
+ * zero.<br>
+ * <br>
+ * This event is fired via the {@link ForgeHooks#onLivingBreathe(LivingEntity, int, int)}.<br>
```
We should mention what bus this fires on

pupnewfster commented on src/main/java/net/minecraftforge/event/entity/living/LivingDrownEvent.java:
```diff
@@ -0,0 +1,46 @@
+ * <br>
+ * This event is fired via the {@link ForgeHooks#onLivingBreathe(LivingEntity, int, int)}.<br>
+ * <br>
+ * {@link #isDrowning()} contains if this entity is drowning.<br>
```
This javadoc should be on the method itself

pupnewfster commented on src/main/java/net/minecraftforge/event/entity/living/LivingDrownEvent.java:
```diff
@@ -0,0 +1,46 @@
+import net.minecraftforge.eventbus.api.Cancelable;
+
+/**
+ * LivingDrownEvent is fired whenever a living entity can't breathe and its air aupply is less than or equal to
```
"aupply" should be "supply"

pupnewfster commented on src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java:
```diff
@@ -288,8 +308,37 @@ public void setCanceled(boolean canceled)
+            }
+        }
+
+        public static enum Action {
```
Unnecessary `static` modifier

Shadows-of-Fire committed:
Major fixes for LivingBreathe/DrownEvent and other misc fixes for upstream code.

* Fixed erroneous LivingEntity.java.patch line.
* Fixed formatting for ForgeHooks.URL_PATTERN.
* Added documentation for LivingBreatheEvent and LivingDrownEvent
* Fixed an issue where the LivingBreatheEvent's implementation caused you to gain air while underwater.
* Fixed an issue where the LivingBreatheEvent's implementation caused entities to not be dismounted in water.
* Expanded the functionality of LivingDrownEvent to allow for controlling the damage amount and particle count.
* Sorted new exc entries.
* Minor style fixes.

Shadows-of-Fire committed:
Format code from upstream to match current style

Also reverts the gradle changes from commit 62317f5f64b2e3ec718a0754ab182dd26788a1e8

Shadows-of-Fire committed:
Further formatting of upstream code

This also applies a full format to ForgeHooks since it was very inconsistent style-wise.

Shadows-of-Fire committed:
Re-implement AlterGroundEvent in a more performant way

This change makes the event fire once per invocation of the AlterGroundDecorator, where it may change the supplied BlockStateProvider (which is then used many times), instead of firing once per altered blockpos.

KostromDan commented:
This broke Grapple Hook mod. Dev saying it's not his fault and won't fix it.
https://github.com/yyon/grapplemod/issues/126
https://github.com/yyon/grapplemod/issues/127
https://github.com/yyon/grapplemod/issues/130
https://github.com/yyon/grapplemod/issues/132
https://github.com/yyon/grapplemod/issues/135

TelepathicGrunt commented:
@KostromDan are you trying to play that mod on 1.20.1 NeoForge? Please use latest Forge instead for 1.20.1. We dropped support for 1.20.1 NeoForge to focus on 1.20.2 and newer since it was too much of a burden to keep backwards compat with Forge on 1.20.1
